building_map_tools/building_crowdsim/navmesh/polygon_factory.py

```diff
@@ -0,0 +1,397 @@
+            # only 1 common_lane expected
+            assert(len(common_lane) == 1)
+            edge = Edge()
+            edge.init_edge(vertices[idx].id,
+                           vertices[idx - 1].id,
+                           polygon.id,
+                           polygon.id)
+            edge.cross_lane_id = list(common_lane)[0]
+            self.edge_manager.add_obj(edge)
+            polygon.edge_ids.add(edge.id)
+
+    def lane_polygon_general_case(self, polygon, already_vertices):
+        """
+        LanePolygon general case does not generate new polygon vertex. This
+        function aims to rearrange the sequence of the 4 added polygon
+        vertices. 0 and 1 are added together; 2 and 3 are added together
+        The idea is checking whether the v0, v2 is in the same side of the lane
+        If in the same side of lane, the sequence is (0, 2, 3, 1)
+        If not, the sequence is (0, 1, 2, 3)
+        """
+        lane = self.lane_manager.data[polygon.related_lane_id]
+        v0 = self.polygon_vertex_manager.data[already_vertices[0]]
+        v2 = self.polygon_vertex_manager.data[already_vertices[2]]
+
+        if self.connection_manager.is_on_same_side_of_lane(lane.id, v0, v2):
+            already_vertices = [
+                already_vertices[0],
+                already_vertices[2],
+                already_vertices[3],
+                already_vertices[1]]
+
+        for v_id in already_vertices:
+            polygon.add_vertex(self.polygon_vertex_manager.data[v_id])
+
+    def lane_polygon_dead_end_case(self, polygon, already_vertices):
+        """
+        LanePolygon dead end case, means this lane is only connected with
+        1 hub polygon. There are only 2 already generated vertices.
+        2 additional dead end polygon vertices should be generated.
+        The sequence of polygon vertices will also be taken care of
+        as general case.
+        """
+        assert(isinstance(polygon, LanePolygon))
+        assert(len(already_vertices) == 2)
+        v0 = self.polygon_vertex_manager.data[already_vertices[0]]
+        v1 = self.polygon_vertex_manager.data[already_vertices[1]]
+
+        lane = self.lane_manager.data[polygon.related_lane_id]
+        construct_vertices = []
+        dead_end_id = -1
+        for dead_end_id in lane.lane_vertex_id:
+            if len(
+                self.lane_vertex_manager
+                    .data[dead_end_id]
+                    .related_lane_ids) == 1:
+                break
+        construct_vertices = self.dead_end_lane_vertices(
+            polygon.id, dead_end_id)
+
+        v2 = construct_vertices[0]
+        v3 = construct_vertices[1]
+        self.polygon_vertex_manager.add_obj(v2)
+        self.polygon_vertex_manager.add_obj(v3)
+
+        if self.connection_manager.is_on_same_side_of_lane(lane.id, v0, v2):
+            result = [v0, v2, v3, v1]
+        else:
+            result = [v0, v1, v2, v3]
```

Comment:
There is a similar rearrangement logic in `lane_polygon_general_case`, can the logic be moved to a shared function?